protos/file.proto

```diff
@@ -78,13 +292,49 @@ message Metadata {
     //   position = stats_page_table[5][0];
     //   length = stats_page_table[5][1];
     // ```
-      uint64 page_table_position = 3;
+    uint64 page_table_position = 3;
   }
 
+  // File statistics.  This field is only present in Lance version 1.
+  //
+  // In Lance version 2 statistics are stored per-column and this field
+  // will not be present.
   StatisticsMetadata statistics = 5;
+
+  // The number of columns in the file (including inner columns when there
+  // are nested fields)
+  //
+  // This can be used to access the column metadata offsets array which is
+  // stored immediately before manifest_position.
+  //
+  // Given N columns the column metadata positions and sizes are stored in a
+  // contiguous buffer of 2*N uint64 values immediately preceding the file
+  // descriptor (or the metadata if the file is not self describing).
+  //
+  // If we let `column_offsets_pos` be:
+  //   manifest_position - (16 * num_columns)
+  //
+  // Then the metadata for column x starts at the uint64:
+  //   file[column_offsets_pos + (16 * x)]
+  // The size of the metadata for column x is given by the uint64:
+  //   file[column_offsets_pos + (16 * x) + 8]
+  //
+  // This field is ignored in Lance version 1 files (the page table is used instead)
+  uint32 num_columns = 6;
```

Comment:
I think this solves problem 1 described in https://github.com/lancedb/lance/issues/1809, but I'm not sure if it solves problem 2. Could you describe how `x` is derived? Do we skip fields with children that have no buffers of their own (struct, fsl in the future)?

Reply:
I think there's two parts:

 * Gaps in field ids: I'm not sure yet why we needed to fill these gaps in the old approach.  It's very possible I'm missing something.  I was thinking, when we read the file, the column metadata at index X corresponds to the field at index X (in dfs-pre-order) of the schema in the file descriptor.  Field IDs are a table concern and not something that the file format bothers with (other than storing and loading them faithfully)
 * Transient columns (structs / lists)
It's probably true that lists will never have a validity buffer (since we can always use negative values as sentinels) but we still need a flag in the metadata telling us if we need to post-process the sentinels or not.  Also, (non-fixed-size) lists do have an offsets buffer, and there are many different ways that buffer could be encoded.

Struct might have a nullability bitmap or they might not.  If the page doesn't have any nulls we can skip the buffer.  If the page has nulls but doesn't have any all-null-structs then we can use a sentinel.  If the page has nulls and all-null-structs then we need a bitmap.

We could also encode structs using a sparse bitmap like they do in procella.  In this case there is a one validity bitmap buffer AND an offsets buffer per child.

The same goes with fixed-size-list.  There may be a need for a validity buffer or we may choose to do some kind of sparse encoding of the FSL (basically turning it into a variable size list where zero-size lists represent null) if most of the data is null.

Reply:
> I'm not sure yet why we needed to fill these gaps in the old approach.

In the old way, the page table is represented in memory as a `BTreeMap<FieldId, (Offset, Length)>`. That `BTreeMap` is written to disk as an array, where the `i`th element has field id `field_ids_offset + i`. So as-is, the page table is coupled to the field ids.

> I was thinking, when we read the file, the column metadata at index X corresponds to the field at index X (in dfs-pre-order) of the schema in the file descriptor.

This is fine, we just need to be really specific about what the dfs-pre-order is. Right now FSL doesn't have child field, and thus the current page table ignores them, even though it does have entries for structs. However, later we want to change schemas to include child fields of FSL, so they are aligned with Arrow schemas. We can either do that now, in our definition of the dfs-pre-order, or find some compatible way to make that change in the future.

I think my main concern at this point is the transition plan for FSL children. The status quo that we have column metadata for lists and structs makes sense.

Reply:
I think I follow your concern better now.

The pragmatic answer to your question is that the only thing that matters for nesting is struct columns.  FSL, regular lists, and maps do not have "depth" in the proposal (the protobuf encoding is recursive and does have depth but this is true even for primitive columns).

The outside-the-box new-age answer is that the file format does not have any depth at all.  There is just a flat list of columns.  "nesting" is an in-memory / arrow concept and we have to map that concept onto the file format in the reader /writer.  For example, we could choose to map arrow struct arrays into a single "packed" column.  We could choose to map arrow struct arrays into a single "shredded" column (the list of children, instead of being at the "column" level, would be instead nested into the encoding).  Both of these approaches are probably bad ideas since they prevent any kind of nested projection pushdown (though the "packed struct" may still be useful in certain cases).

We could even nest other things.  We could create a special kind of FSL, let's call it a tuple, where we encode each index into its own "lance column".  This would allow us to do nested tuple projection.  Although, again, being practical, it would be less work for everyone else (e.g. arrow, etc.) if we implemented such a thing as a struct.